### DISCUSSION TITLE
FEATURES IDEAS

### DISCUSSION THREAD

**Discussion by cedricloneux** [ReagentX/imessage-exporter#98] [Ideas]
Hey,

I was thinking about the app while working on homebrew support and I got a few ideas:

- DOCUMENTATION: document how iMessages and the reverse engineering works for curious people and useful for contributors

- CI/CD: I have seen how the file build.sh is different for dev branch. Simply puting tag with correct names combined with GitHub actions seems easier and more correct to me. The GitHub actions will do the rest

- BACKUP INCREMENTATION: are previous exported backups (html or txt) happened when exporting to a path with existing backup? If not, it would be nice to do so. 
- So people can export sensitive conversations and store in safe place and delete from device. When later they wanna archive in safe place again, they can export and if the output path is existing backup then it’s incremented. That way there is always one complete HTML per contact. It could also be an option. —increment or —fuse 

- MULTIPLE DB AT ONCE: Ability to create a single export from multiple database. People may backup multiple chat.db files. It would be useful to be able to create exports from multiple chat.db at once. So if user backup his conversations db and delete it from device. Then export his convs again later with new messages. He is able to create an export like it’s a single database. That mean still one HTML file per contact 

- INTERNET INDEPENDENCY: if there data such as videos stored on iCloud with url or voice memos working, or YouTube url. Export could download these items as offline. (at least thumbnails for web urls such as youtube). This would remove internet and external servers dependency. Most URL die with time, with this features exports would  keep the same state even years later without any data loss. 

- DATE RANGE OPTION: extract from a date range passed as argument like —before <DATE> —after <DATE> 

- CONTACT OPTION: extract for specific list of contact passed as argument 

- STATS: could have an option to get stats such as word/messages ratios (who talked the most) and other stats about all conversations, a specific conversations or group of conversations. 

- PERSPECTIVE: I think this app could be part of a bigger project. That project would be a group of apps that aims to free people from apple apps keeping their data captive. For instance: export notes, photos (by keeping the albums directories with symlinks etc) also propose to optimise images spaces by converting all of them to lightweight lossless formats such as AVIF and other closed Apple apps.

**Comment by ReagentX** [comment]
Detailed docs on how we interact with the database are here: https://docs.rs/imessage-database/latest/imessage_database/

I don't want this project's scope to grow too large. Backups should be handled by the user, as everyone's situation is different. 

Date range support was added in Lemonade Berry.
